:bug: [client] Start watching for sub-channel

matthiasgeihs opened the pull request:
Closes #251 

Maybe add tests for watching sub-channels. We only test watching ledger channels.

matthiasgeihs commented:
Converting to draft, need to check a few things before we can review and merge.

manoranjith commented on client/init_test.go:
```diff
@@ -24,5 +24,5 @@ import (
 // and blackbox tests (files *_test.go in packages client and client_test).
 func init() {
-	plogrus.Set(logrus.WarnLevel, &logrus.TextFormatter{ForceColors: true})
+	plogrus.Set(logrus.TraceLevel, &logrus.TextFormatter{ForceColors: true})
```
Did you intended to change the logging level permanently or is it only for testing and you missed to revert ?

matthiasgeihs replied:
Intended, will place it in a separate commit.

manoranjith commented on client/test/backend.go:
```diff
@@ -180,6 +181,10 @@ func (b *MockBackend) Withdraw(_ context.Context, req channel.AdjudicatorReq, su
 		log.Debug("withdraw: already concluded:", ch)
+		e := b.latestEvents[ch]
+		if e.Version() != req.Tx.Version {
+			return fmt.Errorf("Invalid version: expted %v, got %v", e.Version(), req.Tx.Version)
```
Should we change `expted` -> `expected` ?

matthiasgeihs replied:
Fixed that already. You commented on an outdated commit.

manoranjith reviewed (commented):
@matthiasgeihs  Changes related to fix the bug in Start Watching for sub-channels and the related tests that  were added LGTM. However, I find the naming conventions were different from the one used for the other tests.
Going by the existing convention, I think parts of code for `Tim` and `Susie` could go into  a separate file and `DisputeSusie`, `DisputeTim` would be just `Susie` and `Tim` (just like `Mallory` and `Carol`). What do you think ?

Also, I was not able to clearly follow the other changes (additional checks for withdraw, register). Are they related to this change ? If yes, could you provide some additional context about these changes in the commit message ?

matthiasgeihs commented on client/adjudicate.go:
```diff
@@ -71,7 +71,14 @@ func (c *Channel) startWatching() (watcher.StatesPub, watcher.AdjudicatorSub, er
+	statesPub, eventsSub, err := func() (watcher.StatesPub, watcher.AdjudicatorSub, error) {
+		if c.IsLedgerChannel() {
+			return c.client.watcher.StartWatchingLedgerChannel(c.Ctx(), signedState)
+		} else {
+			return c.client.watcher.StartWatchingSubChannel(c.Ctx(), c.parent.ID(), signedState)
+		}
```
Using the interface made me think whether it wouldn't be better to just have one interface StartWatching(Context, Channel, Parent)? 🤔 

Anyways, applied your suggestion.

manoranjith replied:
I thought about it, but then what should be the value of parent when we start watching for ledger channel ?

We may not have avoid using `nil` or equivalent, because, this API could be accessed locally or over remote. In case of accessing over remote, the API call could involve one or two programming languages (client stub & server) and serialization for transporting over network.

matthiasgeihs commented:
> @matthiasgeihs Changes related to fix the bug in Start Watching for sub-channels and the related tests that were added LGTM. However, I find the naming conventions were different from the one used for the other tests. Going by the existing convention, I think parts of code for `Tim` and `Susie` could go into a separate file and `DisputeSusie`, `DisputeTim` would be just `Susie` and `Tim` (just like `Mallory` and `Carol`). What do you think ?

We already were using a mixed naming convention. For the role tests, we were using specific first names. The virtual channel test however works differently, for a number of reasons I won't go into here. I found it hard to come up with new names this time, hence the naming I chose. I also think it's more expressive to name the tests by what they do instead of "TestHappySusieTim". We place the code of the two roles in one file together also for the sub-channel happy tests.

> Also, I was not able to clearly follow the other changes (additional checks for withdraw, register). Are they related to this change ? If yes, could you provide some additional context about these changes in the commit message ?

The mock backend has only very limited functionality for testing exactly what we need in the relevant test cases. For making dispute tests work with the mock backend, I had to enhance it's functionality. Before it was mostly used for happy case testing.

matthiasgeihs commented:
Addressed all comments, still playing around with the timeout in the virtual channel dispute tests. I was hoping that I could decrease it a bit, but doesn't seem to be the case...

matthiasgeihs commented:
CI stability issues seem actually somewhat unrelated to the timeout in the virtual channel dispute test. At least it is not the virtual channel tests failing anymore.

I also saw that when the ledger channel watcher is shutting down, it sometimes errors with `Error de-registering channel from watcher: cannot de-register when sub-channels are present`.

matthiasgeihs commented:
Force push:
- Fixed an issue where under certain circumstances an event was written to a closed event channel.
- Fixed an issue where the test expected to receive events for registration of version 0 and version 1, while this may not be true as our event subscription logic may skip outdated events.

matthiasgeihs commented:
The last commit fixes an issue where a panic in one of the role executions was not propagated and thus resulting in a deadlock because the waitHandler was waiting for the client to be closed.

manoranjith commented on client/client_persistence_test.go:
```diff
@@ -27,7 +27,7 @@ func TestPersistencePetraRobert(t *testing.T) {
 	defer cancel()
 
-	runTwoPartyTest(ctx, t, func(rng *rand.Rand) (setups []ctest.RoleSetup, roles [2]ctest.Executer) {
+	runAliceBobTest(ctx, t, func(rng *rand.Rand) (setups []ctest.RoleSetup, roles [2]ctest.Executer) {
```
I think, the new name `runAliceBobTest` is misleading. Because, the tests is actually being run for `Petra` and `Robert` at one place and for `Alice` and `Bob` at another place.

`runTwoPartyTest` or `runTwoPartHappyTest` could be better names for this function.

matthiasgeihs replied:
Reverting this change, because it apparently created other inconsistencies.

manoranjith commented:
@matthiasgeihs LGTM overall. I have two minor suggestions:

1. Naming of `runAliceBobTest`, which I have mentioned above.
2. Reg the commit messages. Could you wrap the body of the commit message at 72 characters ? It is a widely used limit that makes the commit history more readable across different tools.

matthiasgeihs commented:
Rebase

manoranjith reviewed (approved):
LGTM.
